### PR TITLE
Make space for MeshObject

### LINKED ISSUE #12827: Extend Shape classes to support triangle based shapes
This is to support the ability to load in scanned images of sample geometries.

This will require inheriting from object or surface class and providing a class that can support all the same methods but using a triangular based mesh as the source data.

### PR DESCRIPTION
**Pull request by KarlPalmen**
The geometry objects are rearranged to enable a MeshObject to be defined.

IObject class is defined. Object is renamed CGSObject and made to inherit from IObject.
ObjComponent is modified to use IObject instead of CSGObject.

**To test:**

Code Review and you may check that object can be rendered in instrument view or elsewhere.

Fixes #12827, except that MeshObject still needs to be added. The merger of this change beforehand will help with this.

CSGObject has a performance test. Run it (as the Object performance test) without this fix and with this fix to see whether there is a significant loss in performance.

Are you satisfied with the set of methods of IObject. Could any be removed?

**Release Notes** 
Will be added after MeshObject has been added.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/He3TubeEfficiency.cpp [hunk 1]
```diff
@@ -9,7 +9,7 @@
 #include "MantidDataObjects/Workspace2D.h"
 #include "MantidGeometry/Instrument.h"
 #include "MantidGeometry/Instrument/ParameterMap.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (gvardany) [inline Framework/Algorithms/src/He3TubeEfficiency.cpp]: IObject.h

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp [hunk 2]
```diff
@@ -11,7 +11,7 @@
 #include "MantidGeometry/Instrument/DetectorGroup.h"
 #include "MantidGeometry/Instrument/ReferenceFrame.h"
 #include "MantidGeometry/Objects/BoundingBox.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (gvardany) [inline Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: IObject

File: Framework/Algorithms/src/VesuvioL1ThetaResolution.cpp [hunk 3]
```diff
@@ -7,7 +7,7 @@
 #include "MantidAPI/SpectrumInfo.h"
 #include "MantidAPI/WorkspaceFactory.h"
 #include "MantidGeometry/Instrument.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (gvardany) [inline Framework/Algorithms/src/VesuvioL1ThetaResolution.cpp]: IObject

File: Framework/API/inc/MantidAPI/Sample.h [hunk 4]
```diff
@@ -138,7 +138,7 @@ class MANTID_API_DLL Sample {
   /// The sample name
   std::string m_name;
   /// The sample shape object
-  Geometry::Object m_shape;
+  Geometry::CSGObject m_shape;
```
Comment (mantid-roman) [inline Framework/API/inc/MantidAPI/Sample.h]: To accept an `IObject` this should be a pointer (shared?, unique?).

File: Framework/Geometry/inc/MantidGeometry/Instrument/InstrumentDefinitionParser.h [hunk 5]
```diff
@@ -26,7 +26,7 @@ class ICompAssembly;
 class IComponent;
 class Instrument;
 class ObjComponent;
-class Object;
+class CSGObject;
```
Comment (gvardany) [inline Framework/Geometry/inc/MantidGeometry/Instrument/InstrumentDefinitionParser.h]: IObject

File: Framework/Geometry/inc/MantidGeometry/Instrument/InstrumentVisitor.h [hunk 6]
```diff
@@ -24,7 +24,8 @@ class IComponent;
 class IDetector;
 class IObjComponent;
 class Instrument;
-class Object;
+class IObject;
+class CSGObject;
```
Comment (gvardany) [inline Framework/Geometry/inc/MantidGeometry/Instrument/InstrumentVisitor.h]: remove CSGObject?

File: Framework/API/src/SampleShapeValidator.cpp [hunk 7]
```diff
@@ -1,6 +1,6 @@
 #include "MantidAPI/SampleShapeValidator.h"
 #include "MantidAPI/Sample.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (mantid-roman) [inline Framework/API/src/SampleShapeValidator.cpp]: The include is probably unnecessary.

File: Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h [hunk 8]
```diff
@@ -259,14 +266,14 @@ class MANTID_GEOMETRY_DLL Object {
...
 protected:
-  std::vector<const Surface *>
-      SurList; ///< Full surfaces (make a map including complementary object ?)
+  std::vector<const Surface *> m_SurList; ///< Full surfaces (make a map
+  /// including complementary object ?)
 };
 
 /// Typdef for a shared pointer
-typedef boost::shared_ptr<Object> Object_sptr;
+typedef boost::shared_ptr<CSGObject> Object_sptr;
```
Comment (gvardany) [inline Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h]: Shouldn't the typedefs be renamed to CSGObject_sptr?

File: Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h [hunk 9]
```diff
@@ -259,14 +266,14 @@ class MANTID_GEOMETRY_DLL Object {
...
 /// Typdef for a shared pointer to a const object
-typedef boost::shared_ptr<const Object> Object_const_sptr;
+typedef boost::shared_ptr<const CSGObject> Object_const_sptr;
```
Comment (gvardany) [inline Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h]: the same as above

File: Framework/Algorithms/inc/MantidAlgorithms/AbsorptionCorrection.h [hunk 10]
```diff
@@ -113,11 +114,11 @@ class DLLExport AbsorptionCorrection : public API::Algorithm {
    */
   virtual void initialiseCachedDistances() = 0;
 
-  API::MatrixWorkspace_sptr m_inputWS;    ///< A pointer to the input workspace
-  const Geometry::Object *m_sampleObject; ///< Local cache of sample object.
-  Kernel::V3D m_beamDirection;            ///< The direction of the beam.
-  std::vector<double> m_L1s,              ///< Cached L1 distances
-      m_elementVolumes;                   ///< Cached element volumes
+  API::MatrixWorkspace_sptr m_inputWS; ///< A pointer to the input workspace
+  const Geometry::CSGObject *m_sampleObject; ///< Local cache of sample object.
```
Comment (mantid-roman) [inline Framework/Algorithms/inc/MantidAlgorithms/AbsorptionCorrection.h]: Will `AbsorptionCorrection` work with any object (`IObject`) or only with CSG?

File: Framework/Algorithms/inc/MantidAlgorithms/AnnularRingAbsorption.h [hunk 11]
```diff
@@ -12,7 +12,7 @@ class Material;
 class V3D;
 }
 namespace Geometry {
-class Object;
+class CSGObject;
```
Comment (mantid-roman) [inline Framework/Algorithms/inc/MantidAlgorithms/AnnularRingAbsorption.h]: Is this used?

File: Framework/Algorithms/inc/MantidAlgorithms/AnyShapeAbsorption.h [hunk 12]
```diff
@@ -106,7 +107,7 @@ class DLLExport AnyShapeAbsorption : public AbsorptionCorrection {
   std::string sampleXML() override;
   void initialiseCachedDistances() override;
   /// Create the gague volume for the correction
-  Geometry::Object constructGaugeVolume();
+  Geometry::CSGObject constructGaugeVolume();
```
Comment (mantid-roman) [inline Framework/Algorithms/inc/MantidAlgorithms/AnyShapeAbsorption.h]: Will `AnyShapeAbsorption` work with `MeshObjects`?

File: Framework/Geometry/src/Instrument/ComponentHelper.cpp [hunk 13]
```diff
@@ -2,7 +2,7 @@
 #include "MantidGeometry/Instrument/ParameterMap.h"
 #include "MantidGeometry/IComponent.h"
 #include "MantidGeometry/Instrument/ReferenceFrame.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (gvardany) [inline Framework/Geometry/src/Instrument/ComponentHelper.cpp]: IObject.h

File: Framework/Geometry/src/Instrument/ComponentInfo.cpp [hunk 14]
```diff
@@ -1,6 +1,6 @@
 #include "MantidGeometry/Instrument/ComponentInfo.h"
 #include "MantidGeometry/Objects/BoundingBox.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (gvardany) [inline Framework/Geometry/src/Instrument/ComponentInfo.cpp]: IObject.h

File: Framework/Geometry/src/Instrument/ObjComponent.cpp [hunk 15]
```diff
@@ -1,7 +1,7 @@
 #include "MantidGeometry/Instrument/ObjComponent.h"
 #include "MantidGeometry/Instrument/ComponentVisitor.h"
 #include "MantidGeometry/Instrument/ComponentInfo.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (gvardany) [inline Framework/Geometry/src/Instrument/ObjComponent.cpp]: IObject.h

File: Framework/Geometry/src/Rendering/vtkGeometryCacheReader.cpp [hunk 16]
```diff
@@ -7,16 +7,16 @@
...
 #include <Poco/DOM/Element.h>
 
-using Poco::XML::DOMParser;
-using Poco::XML::Document;
-using Poco::XML::Element;
-
 #include "MantidKernel/Exception.h"
 #include "MantidKernel/Logger.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (gvardany) [inline Framework/Geometry/src/Rendering/vtkGeometryCacheReader.cpp]: IObject.h

File: Framework/PythonInterface/mantid/geometry/src/Exports/Object.cpp [hunk 17]
```diff
@@ -1,27 +1,29 @@
 #include "MantidPythonInterface/kernel/GetPointer.h"
```
Comment (gvardany) [inline Framework/PythonInterface/mantid/geometry/src/Exports/Object.cpp]: Should this file be renamed to CSGObject.cpp ?

File: qt/widgets/instrumentview/inc/MantidQtWidgets/InstrumentView/UnwrappedSurface.h [hunk 18]
```diff
@@ -4,6 +4,7 @@
 #include "MantidKernel/V3D.h"
 #include "MantidKernel/Quat.h"
 #include "MantidGeometry/IComponent.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (gvardany) [inline qt/widgets/instrumentview/inc/MantidQtWidgets/InstrumentView/UnwrappedSurface.h]: IObject.h

File: Framework/Algorithms/inc/MantidAlgorithms/SampleCorrections/MCInteractionVolume.h [hunk 19]
```diff
@@ -62,7 +62,7 @@ class MANTID_ALGORITHMS_DLL MCInteractionVolume {
                              double lambdaAfter) const;
 
 private:
-  const Geometry::Object &m_sample;
+  const Geometry::CSGObject &m_sample;
```
Comment (mantid-roman) [inline Framework/Algorithms/inc/MantidAlgorithms/SampleCorrections/MCInteractionVolume.h]: `IObject`?

File: Framework/Algorithms/inc/MantidAlgorithms/SphericalAbsorption.h [hunk 20]
```diff
@@ -98,11 +98,11 @@ class DLLExport SphericalAbsorption : public API::Algorithm {
   int version() const override { return (1); }
 
 protected:
-  API::MatrixWorkspace_sptr m_inputWS;    ///< A pointer to the input workspace
-  const Geometry::Object *m_sampleObject; ///< Local cache of sample object.
-  Kernel::V3D m_beamDirection;            ///< The direction of the beam.
-  std::vector<double> m_L1s,              ///< Cached L1 distances
-      m_elementVolumes;                   ///< Cached element volumes
+  API::MatrixWorkspace_sptr m_inputWS; ///< A pointer to the input workspace
+  const Geometry::CSGObject *m_sampleObject; ///< Local cache of sample object.
```
Comment (mantid-roman) [inline Framework/Algorithms/inc/MantidAlgorithms/SphericalAbsorption.h]: `IObject`?

File: Framework/Algorithms/src/DetectorEfficiencyCor.cpp [hunk 21]
```diff
@@ -9,7 +9,7 @@
 #include "MantidAPI/WorkspaceUnitValidator.h"
 #include "MantidGeometry/Instrument.h"
 #include "MantidGeometry/Instrument/ParameterMap.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (mantid-roman) [inline Framework/Algorithms/src/DetectorEfficiencyCor.cpp]: IObject.h ?

File: Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h [hunk 22]
```diff
@@ -60,76 +62,79 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+  void convertComplement(const std::map<int, CSGObject> &);
 
   virtual void print() const;
   void printTree() const;
 
-  bool isValid(const Kernel::V3D &) const; ///< Check if a point is valid
+  bool
+  isValid(const Kernel::V3D &) const override; ///< Check if a point is valid
   bool isValid(const std::map<int, int> &)
       const; ///< Check if a set of surfaces are valid.
```
Comment (gvardany) [inline Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h]: override?

File: Framework/Algorithms/src/FlatPlateAbsorption.cpp [hunk 23]
```diff
@@ -1,7 +1,7 @@
 #include "MantidAlgorithms/FlatPlateAbsorption.h"
 #include "MantidAPI/MatrixWorkspace.h"
 #include "MantidGeometry/Instrument.h"
-#include "MantidGeometry/Objects/Object.h"
+#include "MantidGeometry/Objects/CSGObject.h"
```
Comment (mantid-roman) [inline Framework/Algorithms/src/FlatPlateAbsorption.cpp]: IObject.h?

File: Framework/Algorithms/src/He3TubeEfficiency.cpp (hunk 1 above)
Comment (mantid-roman) [inline Framework/Algorithms/src/He3TubeEfficiency.cpp]: IObject.h?

File: Framework/Algorithms/src/SofQWNormalisedPolygon.cpp (hunk 2 above)
Comment (mantid-roman) [inline Framework/Algorithms/src/SofQWNormalisedPolygon.cpp]: IObject.h?

**Review by gvardany — CHANGES_REQUESTED** [review]
Rename is mostly clean, except in few instances (see inline). Performance seems to be unaffected.

File: Framework/Algorithms/src/VesuvioL1ThetaResolution.cpp (hunk 3 above)
Comment (mantid-roman) [inline Framework/Algorithms/src/VesuvioL1ThetaResolution.cpp]: IObject.h?

**Review by mantid-roman — DISMISSED** [review]

File: Framework/Algorithms/inc/MantidAlgorithms/AbsorptionCorrection.h (hunk 10 above)
Comment (KarlPalmen) [reply Framework/Algorithms/inc/MantidAlgorithms/AbsorptionCorrection.h]: Not with IObject if it keeps a local cache of sample object.

File: Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h (hunk 22 above)
Comment (KarlPalmen) [reply Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h]: Overrides ``IObject.isValid(const Kernel::V3D &)``.

**Review by KarlPalmen — COMMENTED** [review]
Did the whole thing build OK?

File: Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h (hunk 8 above)
Comment (KarlPalmen) [reply Framework/Geometry/inc/MantidGeometry/Objects/CSGObject.h]: I'm hoping there are none remaining.

File: Framework/Algorithms/inc/MantidAlgorithms/AbsorptionCorrection.h [hunk 24]
```diff
@@ -13,7 +14,7 @@ class Sample;
 }
 namespace Geometry {
 class IDetector;
-class Object;
+class IObject;
```
[the remark this answers is not in the document]
Comment (KarlPalmen) [reply Framework/Algorithms/inc/MantidAlgorithms/AbsorptionCorrection.h]: Why CSGObject?

**Comment by KarlPalmen** [comment]
I've looked at the console log of the RHEL7 build and found it has finished with success.

File: Framework/Geometry/src/Rendering/vtkGeometryCacheReader.cpp (hunk 16 above)
Comment (KarlPalmen) [reply Framework/Geometry/src/Rendering/vtkGeometryCacheReader.cpp]: Needed to prevent an incomplete type error.

**Comment by KarlPalmen** [comment]
@gvardany I've addressed all your comments. Do you want look at this once more?

**Review by gvardany — APPROVED** [review]
Looks good on my part.

**Review by martyngigg — APPROVED** [review]

**Pull request merged**